pkg/datarepair/checker/checker.go

```diff
@@ -217,3 +170,114 @@ func contains(a []string, x string) bool {
+		checker.logger.Debug("no pieces on remote segment")
+		return nil
+	}
+
+	missingPieces, err := checker.overlay.GetMissingPieces(ctx, pieces)
+	if err != nil {
+		checker.logger.Error("error getting missing pieces: ", zap.Error(err))
+		return err
+	}
+
+	monStats.remoteSegmentsChecked++
+	pathElements := storj.SplitPath(path)
+	if len(pathElements) >= 2 && pathElements[1] == "l" {
+		monStats.remoteFilesChecked++
+	}
+
+	numHealthy := int32(len(pieces) - len(missingPieces))
+	redundancy := pointer.Remote.Redundancy
+	// we repair when the number of healthy files is less than or equal to the repair threshold
+	// except for the case when the repair and success thresholds are the same (a case usually seen during testing)
+	if numHealthy >= redundancy.MinReq && numHealthy <= redundancy.RepairThreshold && redundancy.RepairThreshold != redundancy.SuccessThreshold {
+		if len(missingPieces) == 0 {
+			checker.logger.Warn("Missing pieces is zero in checker, but this should be impossible -- bad redundancy scheme.")
+			return nil
+		}
+		monStats.remoteSegmentsNeedingRepair++
+		err = checker.repairQueue.Insert(ctx, &pb.InjuredSegment{
+			Path:       path,
+			LostPieces: missingPieces,
+		})
+		if err != nil {
+			return err
+		}
+
+		// delete always returns nil when something was deleted and also when element didn't exists
+		_ = checker.irrdb.Delete(ctx, []byte(path))
+	} else if numHealthy < redundancy.MinReq {
+		// check to make sure there are at least *4* path elements. the first three
+		// are project, segment, and bucket name, but we want to make sure we're talking
+		// about an actual object, and that there's an object name specified
+		if len(pathElements) >= 4 {
+			project, bucketName, segmentpath := pathElements[0], pathElements[2], pathElements[3]
+			lostSegInfo := storj.JoinPaths(project, bucketName, segmentpath)
+			if contains(monStats.remoteSegmentInfo, lostSegInfo) == false {
+				monStats.remoteSegmentInfo = append(monStats.remoteSegmentInfo, lostSegInfo)
+			}
+		}
+
+		monStats.remoteSegmentsLost++
+		// make an entry in to the irreparable table
+		segmentInfo := &pb.IrreparableSegment{
+			Path:               []byte(path),
+			SegmentDetail:      pointer,
+			LostPieces:         int32(len(missingPieces)),
+			LastRepairAttempt:  time.Now().Unix(),
+			RepairAttemptCount: int64(1),
+		}
+
+		// add the entry if new or update attempt count if already exists
+		err := checker.irrdb.IncrementRepairAttempts(ctx, segmentInfo)
+		if err != nil {
+			checker.logger.Error("error handling irreparable segment to queue: ", zap.Error(err))
+			return err
+		}
+	}
+	return nil
+}
+
+// IrreparableProcess picks items from irrepairabledb and spawns a repair worker
+func (checker *Checker) IrreparableProcess(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	limit := 1
+	var offset int64
+	var monStats durabilityStats
+
+	for {
+		seg, err := checker.irrdb.GetLimited(ctx, limit, offset)
+		if err != nil {
+			return err
```

Comment:
should this loop end with first occurred error?

Reply:
we can return or make the logic complicated, by trying to analyze the error and add lot of error handling code, which we can avoid as this irreparable process runs after very short time and also the Identifyinjuredsegment() process will have taken care of it. I thought about it and found it would be clean and simple to return and come back and do it again when its cycle starts.

Also it the function failed, mostly there is good reason and if we try again , most likely we will see bunch of same errors. 

when there are no more rows, the function returns [], nil, which in the next line handles it as it knows there are no more rows to scan and exists the loop gracefully.